File: voxel_engine/src/camera.cpp
#include "camera.hpp"

Camera::Camera() {
    // position
    this->position = vec3(0.0);
    // horizontal angle : toward -Z
    this->horizontal_angle = 3.14f;
    // vertical angle : 0, look at the horizon
    this->vertical_angle = 0.0f;
    // Initial Field of View
    this->fov = 75.0f;
}

vec3 Camera::get_direction() {
    vec3 direction = vec3(
        cos(this->vertical_angle) * sin(this->horizontal_angle),
        sin(this->vertical_angle),
        cos(this->vertical_angle) * cos(this->horizontal_angle)
    );
    return normalize(direction);
}

vec3 Camera::get_right() {
    return vec3(
        sin(this->horizontal_angle - 3.14f/2.0f),
        0,
        cos(this->horizontal_angle - 3.14f/2.0f)
    );
}

vec3 Camera::get_up() {
    return cross( get_right(), get_direction() );   
}

//moves in direction camera is looking
void Camera::move_toward(vec3 change, bool clip_y) {
    vec3 direction = this->get_direction();
    vec3 right = this->get_right();

    vec3 change_pos = direction * change.x + vec3(0.0, 1.0, 0.0) * change.y + right * change.z;
    if (clip_y) {
        change_pos.y = 0;
    }
    this->position += change_pos;
}

void Camera::rotate(vec2 change) {
    this->horizontal_angle += change.x;
    this->vertical_angle += change.y;

    this->vertical_angle = clamp(this->vertical_angle, -pi<float>()/2 + 0.01f, pi<float>()/2 - 0.01f);
}

mat4 Camera::get_camera_projection_matrix(float aspect_ratio) {
    mat4 projection_matrix = perspective(radians(this->fov), aspect_ratio, 0.1f, 230.0f);
    return projection_matrix;
}

mat4 Camera::get_camera_view_matrix() {
    vec3 direction = this->get_direction();
    vec3 right = this->get_right();
    vec3 up = cross( right, direction );

    mat4 view_matrix = lookAt(
        this->position,           // Camera is here
        this->position+direction, // and looks here : at the same position, plus "direction"
        up                        // Head is up (set to 0,-1,0 to look upside-down)
    );

    return view_matrix;
}

File: voxel_engine/mods/camera.ts
import { degToRad } from './math/Math';
import { Matrix4 } from './math/Matrix4';
import {Vector2} from './math/Vector2';
import {Vector3} from './math/Vector3';
import {print} from './api';

class Camera {
    position: Vector3 = new Vector3(0);
    horizontal_angle: f32 = 3.14;
    vertical_angle: f32 = 0;
    fov: f32 = 75;

    get_direction(): Vector3 {
        let direction: Vector3 = new Vector3(
            NativeMathf.cos(this.vertical_angle) * NativeMathf.sin(this.horizontal_angle),
            NativeMathf.sin(this.vertical_angle),
            NativeMathf.cos(this.vertical_angle) * NativeMathf.cos(this.horizontal_angle)
        );
        return direction.normalize();
    }

    get_right(): Vector3 {
        return new Vector3(
            NativeMathf.sin(this.horizontal_angle - NativeMathf.PI/2.0),
            0,
            NativeMathf.cos(this.horizontal_angle - NativeMathf.PI/2.0)
        );
    }

    get_up(): Vector3 {
        return this.get_right().cross(this.get_direction());
    }

    move_toward(change: Vector3, clip_y: boolean): void {
        let direction = this.get_direction();
        let right = this.get_right();

        let change_pos: Vector3 = direction.multiplyScalar(change.x);
        change_pos.y += change.y;
        change_pos = change_pos.add(right.multiplyScalar(change.z));
        if (clip_y) {
            change_pos.y = 0;
        }
        this.position = this.position.add(change_pos);
    }

    rotate(change: Vector2): void {
        this.horizontal_angle += change.x;
        this.vertical_angle += change.y;
        let min: f32 = -NativeMathf.PI / 2.0 + 0.01;
        let max: f32 = NativeMathf.PI / 2.0 - 0.01;
        if (this.vertical_angle < min) {
            this.vertical_angle = min;
        }
        if (this.vertical_angle > max) {
            this.vertical_angle = max;
        }
    }

    get_camera_projection_matrix(aspect_ratio: f32): Matrix4 {
        let proj: Matrix4 = new Matrix4();
        proj.makePerspective(degToRad(this.fov), aspect_ratio, 0.1, 230);
        return proj;
    }

    get_camera_view_matrix(): Matrix4 {
        let direction: Vector3 = this.get_direction();
        let right: Vector3 = this.get_right();
        let up: Vector3 = right.cross(direction);

        let view: Matrix4 = new Matrix4();
        
        view.lookAt(this.position, direction.add(this.position), up);
        
        return view;
    }
}

export {Camera};
